src/container/container.ts

```diff
@@ -278,89 +399,221 @@ class Container implements interfaces.Container {
         return tempContainer.get<T>(constructorFunction);
     }
 
+    private _preDestroy(constructor: any, instance: any): Promise<void> | void {
+        if (Reflect.hasMetadata(METADATA_KEY.PRE_DESTROY, constructor)) {
+            const data: interfaces.Metadata = Reflect.getMetadata(METADATA_KEY.PRE_DESTROY, constructor);
+
+            return instance[data.value]();
+        }
+    }
+    private _removeModuleHandlers(moduleId:number): void {
+        const handlers = this._moduleActivationStore.remove(moduleId)
+        if (handlers.onActivations.length > 0) {
+            this._activations.removeByCondition(onActivation => handlers.onActivations.indexOf(onActivation) !== -1);
+        }
+        if (handlers.onDeactivations.length > 0) {
+            this._deactivations.removeByCondition(onDeactivation => handlers.onDeactivations.indexOf(onDeactivation) !== -1);
+        }
```

Comment:
We can improve the performance of this operation, on my way to refactor de activations store

Reply:
Sure go ahead.

Reply:
Instead of putting the module activations in with those added directly from the container, keep them separate and merge them together when required.

e.g have a method on the container instead of below
```
//resolver.ts
const _getContainerActivationsForService = <T>(container: interfaces.Container, serviceIdentifier: interfaces.ServiceIdentifier<T>) => {
    // smell accessing _activations, but similar pattern is done in planner.getBindingDictionary()
    const activations = (container as any)._activations as interfaces.Lookup<interfaces.BindingActivation<any>>;

    return activations.hasKey(serviceIdentifier) ? activations.get(serviceIdentifier).values() : [].values();
}
//container.ts
private _deactivate<T>(binding: Binding<T>, instance: T): void | Promise<void> {
        const constructor = Object.getPrototypeOf(instance).constructor;

        try {
            if (this._deactivations.hasKey(binding.serviceIdentifier)) {
                const result = this._deactivateContainer(
                    instance,
                    this._deactivations.get(binding.serviceIdentifier).values(),
                );
```

When unload no need for _removeModuleHandlers logic - just remove the module bindings from the moduleActivationStore
```
public unload(...modules: interfaces.ContainerModule[]): void {
        modules.forEach((module) => {
            const deactivations = this._removeModuleBindings(module.id)
            this._deactivateSingletons(deactivations);

            this._removeModuleHandlers(module.id);
        });

    }

```

Reply:
Although this does not retain activation order